gio/src/list_model.rs

```diff
@@ -0,0 +1,88 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use std::cell::Cell;
+use std::marker::PhantomData;
+use std::rc::Rc;
+
+use glib::{clone, IsA, Object};
+
+use crate::prelude::*;
+use crate::ListModel;
+
+pub trait ListModelExtManual: ListModelExt {
+    /// Returns an iterator over the ListModel elements.
+    fn iter<T: IsA<Object>>(&self) -> ListModelIter<Self, T>
+    where
+        Self: Sized + Clone;
+}
+
+impl<O: IsA<ListModel>> ListModelExtManual for O {
+    fn iter<T: IsA<Object>>(&self) -> ListModelIter<Self, T>
+    where
+        Self: Sized + Clone,
+    {
+        ListModelIter::new(self.clone())
+    }
+}
+
+pub struct ListModelIter<M: ListModelExt, T: ObjectExt> {
+    data_type: PhantomData<T>,
+    model: M,
+    i: Rc<Cell<u32>>,
+}
+
+impl<M: ListModelExt + Clone, T: IsA<Object>> ListModelIter<M, T> {
+    /// Creates a new ListModel iterator.
+    /// 
+    /// ### Panics
+    /// Panics if `T::static_type()` is not of the model’s item type.
+    fn new(model: M) -> Self {
+        if !T::static_type().is_a(model.item_type()) {
```

Comment:
This is probably inversed. If you have a model of widgets, you don't want to be able to iterate over it as buttons (that might fail), but if you have a model of buttons you want to be able to iterate over it as widgets (that can't fail).

Reply:
That’s a bad one. You’re right, it should be contravariant. Thank you for taking the time *really* looking into this.